rust/src/normalization.rs

```diff
@@ -45,3 +49,496 @@ pub(crate) fn normalize_etc_shadow(rootfs: &openat::Dir) -> Result<()> {
+            Err(e) => return Err(e.into()),
+            _ => (),
+        };
+
+        // Make sure things are sane
+        if buffer[..8] != RPM_HEADER_MAGIC {
+            return Err(anyhow!("Bad RPM header magic in RPM database"));
+        }
+
+        // Grab the count of index records and the size of the data blob
+        let record_count = u32::from_be_bytes(buffer[8..12].try_into()?);
+        let data_size = u32::from_be_bytes(buffer[12..].try_into()?);
+
+        // Loop through the records looking for ones that point at things
+        // that are, or are derived from, timestamps
+        let mut offsets = Vec::new();
+        for _ in 0..record_count {
+            rpmdb.read_exact(&mut buffer)?;
+
+            let tag = u32::from_be_bytes(buffer[..4].try_into()?);
+            if tag == RPMTAG_INSTALLTIME || tag == RPMTAG_INSTALLTID {
+                offsets.push((tag, u32::from_be_bytes(buffer[8..12].try_into()?)));
+            }
+        }
+
+        // Work through the data blob replacing the timestamp-derived values
+        // with the timestamp we want
+        offsets.sort_unstable_by_key(|(_, offset)| *offset);
+        let mut offset = 0;
+        for (tag, value_offset) in offsets {
+            rpmdb.seek(std::io::SeekFrom::Current((value_offset - offset) as i64))?;
+            if tag == RPMTAG_INSTALLTID {
+                rpmdb.write_all(&install_tid.to_be_bytes())?;
+            } else if tag == RPMTAG_INSTALLTIME {
+                rpmdb.write_all(&install_time.to_be_bytes())?;
+                install_time += 1;
+            }
+            offset = value_offset + std::mem::size_of::<u32>() as u32;
+        }
+
+        // Move to the next record
+        rpmdb.seek(std::io::SeekFrom::Current((data_size - offset) as i64))?;
+    }
+
+    // Seek back to where we were before
+    rpmdb.seek(std::io::SeekFrom::Start(pos))?;
+
+    Ok(())
+}
+
+#[context("Rewriting rpmdb database files for build stability")]
+pub(crate) fn normalize_rpmdb(rootfs: &openat::Dir, rpmdb_path: impl AsRef<Path>) -> Result<()> {
+    let source_date = if let Some(source_date) = source_date_epoch() {
+        source_date as u32
+    } else {
+        return Ok(());
+    };
+
+    let mut bwrap =
+        Bubblewrap::new_with_mutability(rootfs, crate::ffi::BubblewrapMutability::Immutable)?;
+    bwrap.append_child_argv(&["rpm", "--eval", "%{_db_backend}"]);
+    let cancellable = gio::Cancellable::new();
+    let db_backend = bwrap.run_captured(Some(&cancellable))?;
+
+    let db_backend = String::from_utf8(db_backend.to_vec())?;
+
+    match db_backend.trim() {
+        "bdb" => bdb_normalize::normalize(rootfs, rpmdb_path, source_date),
+        "ndb" => Ok(()),
+        "sqlite" => Ok(()),
+        _ => Err(anyhow!("Unknown rpmdb backend: {}", db_backend)),
+    }
+}
+
+mod bdb_normalize {
+    // Gather round, my friends, and I shall tell you a tale of trade-offs and consequences.
+    //
```

Comment:
I love this :smile:

Reply:
The only thing worse than no comments or useless comments is boring comments. 🙃